### PR TITLE
Implemented clDice in args.py and loss.py

### PR DESCRIPTION
**Pull request by braykzmi**
Implemented clDice from jocpae's github : https://github.com/jocpae/clDice/tree/master

Changes made in /mist/runtime/args.py and /mist/runtime/loss.py

### CODE REVIEW

File: mist/runtime/loss.py [hunk 1]
```diff
@@ -31,6 +31,64 @@ def forward(self, y_true, y_pred):
...
+            y_pred = y_pred[:, 1:, :, :]
+        skel_pred = self.soft_skeletonize(y_pred)
+        skel_true = self.soft_skeletonize(y_true)
+        tprec = (torch.sum(torch.multiply(skel_pred, y_true))+self.smooth)/(torch.sum(skel_pred)+self.smooth)    
+        tsens = (torch.sum(torch.multiply(skel_true, y_pred))+self.smooth)/(torch.sum(skel_true)+self.smooth)    
+        cl_dice = 1.- 2.0*(tprec*tsens)/(tprec+tsens)
+        return cl_dice
+
+
+def soft_dice(y_true, y_pred):
```
Comment (aecelaya) [inline mist/runtime/loss.py:55]: This is already implemented in the same file:

https://github.com/mist-medical/MIST/blob/41c6fef4aa3d4d4418023d56e4066c0d227667bd/mist/runtime/loss.py#L15

Please see if you can use this implementation. We don't want to reimplement the same functions.

File: mist/runtime/loss.py [hunk 2]
```diff
@@ -31,6 +31,64 @@ def forward(self, y_true, y_pred):
...
+    Returns:
+        [float32]: [loss value]
+    """
+    smooth = 1
+    intersection = torch.sum((y_true * y_pred))
+    coeff = (2. *  intersection + smooth) / (torch.sum(y_true) + torch.sum(y_pred) + smooth)
+    return (1. - coeff)
+
+
+class soft_dice_cldice(nn.Module):
```
Comment (aecelaya) [inline mist/runtime/loss.py:71]: Use capital letters to name classes. For example, this should be SoftDiceCldice.

File: mist/runtime/loss.py [hunk 3]
```diff
@@ -31,6 +31,64 @@ def forward(self, y_true, y_pred):
...
+    def __init__(self, iter_=3, alpha=0.5, smooth = 1., exclude_background=False):
+        super(soft_dice_cldice, self).__init__()
+        self.iter = iter_
+        self.smooth = smooth
+        self.alpha = alpha
```
Comment (aecelaya) [inline mist/runtime/loss.py:76]: Other loss functions like the GSL, Boundary Loss, and HD loss use an alpha parameter.

Take a look at this page:
https://mist-medical.readthedocs.io/en/latest/advanced_topics/#weighting-schedules-for-boundary-based-loss-functions

It would be nice to make this implementation compatible with the existing alpha schedule implementations.

File: mist/runtime/loss.py [hunk 4]
```diff
@@ -31,6 +31,64 @@ def forward(self, y_true, y_pred):
         loss = torch.mean(loss)
         return loss
 
+#Implement clDice:
+class soft_cldice(nn.Module):
+    def __init__(self, iter_=3, smooth = 1., exclude_background=False):
```
Comment (aecelaya) [inline mist/runtime/loss.py:36]: Be consistent about your spacing with equal signs. As a general rule of thumb, I don't use spaces in arguments and use them when defining or setting values. For example,

def __init__(self, iter_=3, smooth=1):
  self.iter = iter_

Also, why do you use a variable name iter_? Maybe make this clearer with more descriptive variable names like iterations?

File: mist/runtime/loss.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import torch
 import torch.nn as nn
 from torch.nn.functional import softmax, one_hot
-
+from .soft_skeleton import SoftSkeletonize
```
Comment (aecelaya) [inline mist/runtime/loss.py:4]: This should be the following

from mist.runtime.soft_skeleton import SoftSkeletonize

File: mist/runtime/soft_skeleton.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+
+import torch
+import torch.nn.functional as F
```
Comment (aecelaya) [inline mist/runtime/soft_skeleton.py:3]: For the sake of consistency, maybe add 

import torch.nn as nn

Also please rename this file to something like loss_utils.py.

File: mist/runtime/soft_skeleton.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+class SoftSkeletonize(torch.nn.Module):
+
+    def __init__(self, num_iter=40):
+
+        super(SoftSkeletonize, self).__init__()
+        self.num_iter = num_iter
+
+    def soft_erode(self, img):
+
+        if len(img.shape)==4:
```
Comment (aecelaya) [inline mist/runtime/soft_skeleton.py:14]: We won't ever encounter this case since MIST only deals with 3D images. I'd remove this. Also add an else statement that throws an error if len(img.shape) does not equal 5. Something like:

else:
  raise ValueError("[Loss Error] len(img.shape) is not equal to 5")

File: mist/runtime/loss.py [hunk 8]
```diff
@@ -31,6 +31,64 @@ def forward(self, y_true, y_pred):
...
+#Implement clDice:
+class soft_cldice(nn.Module):
+    def __init__(self, iter_=3, smooth = 1., exclude_background=False):
+        super(soft_cldice, self).__init__()
+        self.iter = iter_
+        self.smooth = smooth
+        self.soft_skeletonize = SoftSkeletonize(num_iter=10)
+        self.exclude_background = exclude_background
+
+    def forward(self, y_true, y_pred):
```
Comment (aecelaya) [inline mist/runtime/loss.py:43]: For readability, please space this code out and add comments saying what is going on in this function.

**Comment by aecelaya** [comment]
Thanks for submitting a PR! It's looking good so far. I left some comments we need to address before merging the PR. Feel free to let me know if you have any questions. I'm happy to discuss these further.

**Comment by braykzmi** [comment]
No problem. I will get this fixed tomorrow morning. Thanks!

Warm Regards,
Brayden Mi

File: mist/runtime/loss.py [hunk 9]
```diff
@@ -31,6 +31,64 @@ def forward(self, y_true, y_pred):
...
+    def __init__(self, iter_=3, alpha=0.5, smooth = 1., exclude_background=False):
+        super(soft_dice_cldice, self).__init__()
+        self.iter = iter_
+        self.smooth = smooth
+        self.alpha = alpha
+        self.soft_skeletonize = SoftSkeletonize(num_iter=10)
+        self.exclude_background = exclude_background
+
+    def forward(self, y_true, y_pred):
+        if self.exclude_background:
```
Comment (aecelaya) [inline mist/runtime/loss.py:81]: Also, add these lines here:

```
# Prepare inputs
y_true = get_one_hot(y_true, y_pred.shape[1])
y_pred = softmax(y_pred, dim=1)
```

We apply `get_one_hot` because the data loader does not stream the masks in one hot encoder and the networks don't apply a final activation to their outputs. This latter part gives us more flexibility for future applications like regression instead of segmentation.

File: mist/runtime/loss.py (hunk 8 above)
Comment (aecelaya) [reply mist/runtime/loss.py:43]: Same comment here:

Also, add these lines here:

```
# Prepare inputs
y_true = get_one_hot(y_true, y_pred.shape[1])
y_pred = softmax(y_pred, dim=1)
```

We apply `get_one_hot` because the data loader does not stream the masks in one hot encoder and the networks don't apply a final activation to their outputs. This latter part gives us more flexibility for future applications like regression instead of segmentation.

File: mist/runtime/loss.py (hunk 3 above)
Comment (braykzmi) [reply mist/runtime/loss.py:76]: I'm a little confused on alpha schedule. How should I go about implementing alpha scheduling?

Comment (aecelaya) [reply mist/runtime/loss.py:76]: It's already implemented for you. Take a look here:

https://github.com/mist-medical/MIST/blob/de47168692118afb1910cffd9c86b52d20de9763/mist/runtime/utils.py#L645

And also here:

https://github.com/mist-medical/MIST/blob/de47168692118afb1910cffd9c86b52d20de9763/mist/runtime/run.py#L105

It's not a big deal. I'd say finish this PR first. I can show you how to do this later.

File: mist/runtime/loss.py (hunk 9 above)
Comment (aecelaya) [reply mist/runtime/loss.py:74]: Is this using the same implementation as the ```SoftCLDice```? If so, then create an instance of ```SoftCLDice``` in ```__init``` and call it in the forward pass of this loss. Something like this should do the same thing right? It's always a best practice to reuse your existing implementations if you can. In this case, try reusing your ```SoftCLDice```. Other than this, everything else looks great and we should be ready to merge!

```
class SoftDiceCLDice(nn.Module):
        def __init__(self, iterations=3, alpha=0.5, smooth=1., exclude_background=False):
            super(SoftDiceCLDice, self).__init__()
            self.iterations = iterations
            self.smooth = smooth
            self.alpha = alpha
            self.dice_loss = DiceLoss()
            self.exclude_background = exclude_background
            self.cldice_loss = SoftCLDice(self.iterations, self.smooth, self.exclude_background)

       def forward(self, y_true, y_pred):
           dice = self.dice_loss(y_true, y_pred)
           cldice = self.cldice_loss(y_true, y_pred)
           return (1.0 - self.alpha) * dice + self.alpha * cl_dice
```

**Comment by braykzmi** [comment]
Working on this now.

**Comment by aecelaya** [comment]
Looks great! I'll merge this today. We'll need to test this on a dataset before we release it in the next version

**Review by aecelaya — APPROVED** [review]
